[TIR][Printer] text format printer considering future parsing use

spectrometerHBH opened the pull request:
## Several Points
1. Rename vars/size_vars/buffers to use name hints to identify them
2. A node is either in META or not. If ref A is put in META, ref B is contained in A, then when we need to print ref B we print its META.
3. Now only the `node` in `AttrStmt` may be put in META except that `node` is `StringImm/Var/Buffer/IterVar`.
4. Print the Type of vars when we first encounter them on the tree.
5. This PR hasn't combine relay's printer with tir's printer. I put several choices here for combined API for discussion.
- add an `astext` member function for IRModule: we only do mix printing when printing an IRModule. Keep `relay.astext` and `tir.astext` separate for other uses.
- merge relay's printer & tir's printer into one printer and use only one API `ir.astext`

cc @tqchen @Hzfengsy 

## Several Examples
### Simple
```c++
primfn(A0_1: handle, A1_1: handle, C_1: handle) -> ()
  attr = {"tir.noalias": True, "global_symbol": "main"}
  buffers = {A1: Buffer(A1_2: handle, float32, [m: int32, n: int32], [stride: int32, stride_1: int32], type="auto"),
             A0: Buffer(A0_2: handle, float32, [m, n], [stride_2: int32, stride_3: int32], type="auto"),
             C: Buffer(C_2: handle, float32, [m, n], [stride_4: int32, stride_5: int32], type="auto")}
  buffer_map = {A0_1: A0, C_1: C, A1_1: A1} {
  attr [B.v0: handle] "storage_scope" = "global";
  allocate(B.v0, float32, [n]);
  attr [B.v1: handle] "storage_scope" = "global";
  allocate(B.v1, float32, [n]);
  for (i: int32, 0, m) {
    for (j: int32, 0, n) {
      B.v0[j] = ((float32*)A0_2[((i*stride_2) + (j*stride_3))]) + 2f32)
      B.v1[j] = ((float32*)A0_2[((i*stride_2) + (j*stride_3))])*3f32)
    }
    for (j_1: int32, 0, n) {
      C_2[((i*stride_4) + (j_1*stride_5))] = ((float32*)A1_2[((i*stride) + (j_1*stride_1))]) + (float32*)B.v0[j_1]))
    }
  }
}
```
### Conv on GPU
```c++
primfn(A_1: handle, W_1: handle, B_1: handle) -> ()
  attr = {"tir.noalias": True, "global_symbol": "main"}
  buffers = {W: Buffer(W_2: handle, float32, [3, 3, 256, 512], []),
             B: Buffer(B_2: handle, float32, [14, 14, 512, 256], []),
             A: Buffer(A_2: handle, float32, [14, 14, 256, 256], [])}
  buffer_map = {B_1: B, A_1: A, W_1: W} {
  attr [IterVar(blockIdx.z: int32, (nullptr), "ThreadIndex", "blockIdx.z")] "thread_extent" = 196;
  attr [B.local: handle] "storage_scope" = "local";
  allocate(B.local, float32, [64]);
  attr [Apad.shared: handle] "storage_scope" = "shared";
  allocate(Apad.shared, float32, [512]);
  attr [W.shared: handle] "storage_scope" = "shared";
  allocate(W.shared, float32, [512]);
  attr [Apad.shared.local: handle] "storage_scope" = "local";
  allocate(Apad.shared.local, float32, [8]);
  attr [W.shared.local: handle] "storage_scope" = "local";
  allocate(W.shared.local, float32, [8]);
  attr [IterVar(blockIdx.y: int32, (nullptr), "ThreadIndex", "blockIdx.y")] "thread_extent" = 8;
  attr [IterVar(blockIdx.x: int32, (nullptr), "ThreadIndex", "blockIdx.x")] "thread_extent" = 4;
  attr [IterVar(threadIdx.y: int32, [0:8], "ThreadIndex", "threadIdx.y")] "thread_extent" = 8;
  attr [IterVar(threadIdx.x: int32, [0:8], "ThreadIndex", "threadIdx.x")] "thread_extent" = 8 {
    for (ff.c.init: int32, 0, 4) {
      for (nn.c.init: int32, 0, 4) {
        B.local[((ff.c.init*4) + nn.c.init)] = 0f32
        B.local[(((ff.c.init*4) + nn.c.init) + 32)] = 0f32
        B.local[(((ff.c.init*4) + nn.c.init) + 16)] = 0f32
        B.local[(((ff.c.init*4) + nn.c.init) + 48)] = 0f32
      }
    }
    for (rc.outer: int32, 0, 32) {
      for (ry: int32, 0, 3) {
        for (rx: int32, 0, 3) {
          for (ax3.inner.outer: int32, 0, 2) {
            Apad.shared[ramp((((threadIdx.y*64) + (threadIdx.x*8)) + (ax3.inner.outer*4)), 1, 4)] = @tvm_if_then_else(((((1 <= (floordiv(blockIdx.z, 14) + ry)) and ((floordiv(blockIdx.z, 14) + ry) < 15)) and (1 <= (rx + floormod(blockIdx.z, 14)))) and ((rx + floormod(blockIdx.z, 14)) < 15)), (float32x4*)A_2[ramp((((((((((ry*917504) + (blockIdx.z*65536)) + (rx*65536)) + (rc.outer*2048)) + (threadIdx.y*256)) + (blockIdx.x*64)) + (threadIdx.x*8)) + (ax3.inner.outer*4)) - 983040), 1, 4)]), broadcast(0f32, 4), dtype=float32x4, type="pure_intrin", index=0)
          }
          for (ax3.inner.outer_1: int32, 0, 2) {
            W.shared[ramp((((threadIdx.y*64) + (threadIdx.x*8)) + (ax3.inner.outer_1*4)), 1, 4)] = (float32x4*)W_2[ramp((((((((ry*393216) + (rx*131072)) + (rc.outer*4096)) + (threadIdx.y*512)) + (blockIdx.y*64)) + (threadIdx.x*8)) + (ax3.inner.outer_1*4)), 1, 4)])
          }
          for (rc.inner: int32, 0, 8) {
            for (ax3: int32, 0, 4) {
              Apad.shared.local[ax3] = (float32*)Apad.shared[(((rc.inner*64) + (threadIdx.x*4)) + ax3)])
              Apad.shared.local[(ax3 + 4)] = (float32*)Apad.shared[((((rc.inner*64) + (threadIdx.x*4)) + ax3) + 32)])
            }
            for (ax3_1: int32, 0, 4) {
              W.shared.local[ax3_1] = (float32*)W.shared[(((rc.inner*64) + (threadIdx.y*4)) + ax3_1)])
              W.shared.local[(ax3_1 + 4)] = (float32*)W.shared[((((rc.inner*64) + (threadIdx.y*4)) + ax3_1) + 32)])
            }
            for (ff.c: int32, 0, 4) {
              for (nn.c: int32, 0, 4) {
                B.local[((ff.c*4) + nn.c)] = ((float32*)B.local[((ff.c*4) + nn.c)]) + ((float32*)Apad.shared.local[nn.c])*(float32*)W.shared.local[ff.c])))
                B.local[(((ff.c*4) + nn.c) + 32)] = ((float32*)B.local[(((ff.c*4) + nn.c) + 32)]) + ((float32*)Apad.shared.local[nn.c])*(float32*)W.shared.local[(ff.c + 4)])))
                B.local[(((ff.c*4) + nn.c) + 16)] = ((float32*)B.local[(((ff.c*4) + nn.c) + 16)]) + ((float32*)Apad.shared.local[(nn.c + 4)])*(float32*)W.shared.local[ff.c])))
                B.local[(((ff.c*4) + nn.c) + 48)] = ((float32*)B.local[(((ff.c*4) + nn.c) + 48)]) + ((float32*)Apad.shared.local[(nn.c + 4)])*(float32*)W.shared.local[(ff.c + 4)])))
              }
            }
          }
        }
      }
    }
    for (ff.inner.inner.inner: int32, 0, 4) {
      for (nn.inner.inner.inner: int32, 0, 4) {
        B_2[(((((((blockIdx.z*131072) + (blockIdx.y*16384)) + (threadIdx.y*1024)) + (ff.inner.inner.inner*256)) + (blockIdx.x*64)) + (threadIdx.x*4)) + nn.inner.inner.inner)] = (float32*)B.local[((ff.inner.inner.inner*4) + nn.inner.inner.inner)])
        B_2[((((((((blockIdx.z*131072) + (blockIdx.y*16384)) + (threadIdx.y*1024)) + (ff.inner.inner.inner*256)) + (blockIdx.x*64)) + (threadIdx.x*4)) + nn.inner.inner.inner) + 8192)] = (float32*)B.local[(((ff.inner.inner.inner*4) + nn.inner.inner.inner) + 32)])
        B_2[((((((((blockIdx.z*131072) + (blockIdx.y*16384)) + (threadIdx.y*1024)) + (ff.inner.inner.inner*256)) + (blockIdx.x*64)) + (threadIdx.x*4)) + nn.inner.inner.inner) + 32)] = (float32*)B.local[(((ff.inner.inner.inner*4) + nn.inner.inner.inner) + 16)])
        B_2[((((((((blockIdx.z*131072) + (blockIdx.y*16384)) + (threadIdx.y*1024)) + (ff.inner.inner.inner*256)) + (blockIdx.x*64)) + (threadIdx.x*4)) + nn.inner.inner.inner) + 8224)] = (float32*)B.local[(((ff.inner.inner.inner*4) + nn.inner.inner.inner) + 48)])
      }
    }
  }
}
```
### TensorCore for Conv
```c++
primfn(A_1: handle, W_1: handle, Conv_1: handle) -> ()
  attr = {"tir.noalias": True, "global_symbol": "main"}
  buffers = {W: Buffer(W_2: handle, float16, [3, 3, 16, 32, 16, 16], []),
             Conv: Buffer(Conv_2: handle, float32, [16, 14, 14, 32, 16, 16], []),
             A: Buffer(A_2: handle, float16, [16, 14, 14, 16, 16, 16], [])}
  buffer_map = {Conv_1: Conv, A_1: A, W_1: W} {
  attr [IterVar(blockIdx.z: int32, (nullptr), "ThreadIndex", "blockIdx.z")] "thread_extent" = 196;
  attr [Conv.wmma.accumulator: handle] "storage_scope" = "wmma.accumulator";
  allocate(Conv.wmma.accumulator, float32, [2048]);
  attr [Apad.shared: handle] "storage_scope" = "shared";
  allocate(Apad.shared, float16, [12288]);
  attr [W.shared: handle] "storage_scope" = "shared";
  allocate(W.shared, float16, [12288]);
  attr [Apad.shared.wmma.matrix_a: handle] "storage_scope" = "wmma.matrix_a";
  allocate(Apad.shared.wmma.matrix_a, float16, [512]);
  attr [W.shared.wmma.matrix_b: handle] "storage_scope" = "wmma.matrix_b";
  allocate(W.shared.wmma.matrix_b, float16, [1024]);
  attr [IterVar(blockIdx.x: int32, (nullptr), "ThreadIndex", "blockIdx.x")] "thread_extent" = 2;
  attr [IterVar(blockIdx.y: int32, (nullptr), "ThreadIndex", "blockIdx.y")] "thread_extent" = 4;
  attr [IterVar(threadIdx.y: int32, (nullptr), "ThreadIndex", "threadIdx.y")] "thread_extent" = 4;
  attr [IterVar(threadIdx.z: int32, (nullptr), "ThreadIndex", "threadIdx.z")] "thread_extent" = 2 {
    for (n.c.init: int32, 0, 2) {
      for (o.c.init: int32, 0, 4) {
        @tvm_fill_fragment(Conv.wmma.accumulator, 16, 16, 16, ((n.c.init*4) + o.c.init), 0f32, dtype=handle, type="intrin", index=0)
      }
    }
    for (ic.outer: int32, 0, 8) {
      for (kh: int32, 0, 3) {
        for (ax2: int32, 0, 3) {
          for (ax3: int32, 0, 2) {
            for (ax4.ax5.fused.outer: int32, 0, 8) {
              attr [IterVar(threadIdx.x: int32, (nullptr), "ThreadIndex", "threadIdx.x")] "thread_extent" = 32;
              Apad.shared[((((((threadIdx.y*3072) + (threadIdx.z*1536)) + (ax2*512)) + (ax3*256)) + (ax4.ax5.fused.outer*32)) + threadIdx.x)] = @tvm_if_then_else(((((1 <= (floordiv(blockIdx.z, 14) + kh)) and ((floordiv(blockIdx.z, 14) + kh) < 15)) and (1 <= (ax2 + floormod(blockIdx.z, 14)))) and ((ax2 + floormod(blockIdx.z, 14)) < 15)), (float16*)A_2[(((((((((((blockIdx.x*6422528) + (threadIdx.y*1605632)) + (threadIdx.z*802816)) + (kh*57344)) + (blockIdx.z*4096)) + (ax2*4096)) + (ic.outer*512)) + (ax3*256)) + (ax4.ax5.fused.outer*32)) + threadIdx.x) - 61440)]), 0f16, dtype=float16, type="pure_intrin", index=0)
            }
          }
        }
        for (ax1: int32, 0, 3) {
          for (ax2_1: int32, 0, 2) {
            attr [IterVar(threadIdx.x, (nullptr), "ThreadIndex", "threadIdx.x")] "thread_extent" = 32;
            W.shared[ramp((((((ax1*4096) + (ax2_1*2048)) + (threadIdx.y*512)) + (threadIdx.z*256)) + (threadIdx.x*8)), 1, 8)] = (float16x8*)W_2[ramp(((((((((kh*393216) + (ax1*131072)) + (ic.outer*16384)) + (ax2_1*8192)) + (blockIdx.y*2048)) + (threadIdx.y*512)) + (threadIdx.z*256)) + (threadIdx.x*8)), 1, 8)])
          }
        }
        for (ic.inner: int32, 0, 2) {
          for (kw: int32, 0, 3) {
            for (ax0: int32, 0, 2) {
              @tvm_load_matrix_sync(Apad.shared.wmma.matrix_a, 16, 16, 16, ax0, @tvm_access_ptr(@type_annotation(, dtype=float16, type="pure_intrin", index=0), Apad.shared, ((((threadIdx.y*3072) + (ax0*1536)) + (kw*512)) + (ic.inner*256)), 256, 1, dtype=handle, type="intrin", index=0), 16, "row_major", dtype=handle, type="intrin", index=0)
            }
            for (ax3_1: int32, 0, 4) {
              @tvm_load_matrix_sync(W.shared.wmma.matrix_b, 16, 16, 16, ax3_1, @tvm_access_ptr(@type_annotation(, dtype=float16, type="pure_intrin", index=0), W.shared, ((((kw*4096) + (ic.inner*2048)) + (threadIdx.z*1024)) + (ax3_1*256)), 256, 1, dtype=handle, type="intrin", index=0), 16, "row_major", dtype=handle, type="intrin", index=0)
            }
            for (n.c: int32, 0, 2) {
              for (o.c: int32, 0, 4) {
                @tvm_mma_sync(Conv.wmma.accumulator, ((n.c*4) + o.c), Apad.shared.wmma.matrix_a, n.c, W.shared.wmma.matrix_b, o.c, Conv.wmma.accumulator, ((n.c*4) + o.c), dtype=handle, type="intrin", index=0)
              }
            }
          }
        }
      }
    }
    for (n.inner: int32, 0, 2) {
      for (o.inner: int32, 0, 4) {
        @tvm_store_matrix_sync(Conv.wmma.accumulator, 16, 16, 16, ((n.inner*4) + o.inner), @tvm_access_ptr(@type_annotation(, dtype=float32, type="pure_intrin", index=0), Conv_2, (((((((blockIdx.x*12845056) + (threadIdx.y*3211264)) + (n.inner*1605632)) + (blockIdx.z*8192)) + (blockIdx.y*2048)) + (threadIdx.z*1024)) + (o.inner*256)), 256, 2, dtype=handle, type="intrin", index=0), 16, "row_major", dtype=handle, type="intrin", index=0)
      }
    }
  }
}
```
### GEMM on CPU
```c++
primfn(A_1: handle, B_1: handle, C_1: handle) -> ()
  attr = {"tir.noalias": True, "global_symbol": "main"}
  buffers = {B: Buffer(B_2: handle, float32, [1024, 1024], []),
             C: Buffer(C_2: handle, float32, [1024, 1024], []),
             A: Buffer(A_2: handle, float32, [1024, 1024], [])}
  buffer_map = {C_1: C, A_1: A, B_1: B} {
  attr [packedB: handle] "storage_scope" = "global";
  allocate(packedB, float32x32, [32768]) {
    for (x: int32, 0, 32) "parallel" {
      for (y: int32, 0, 1024) {
        packedB[ramp(((x*32768) + (y*32)), 1, 32)] = (float32x32*)B_2[ramp(((y*1024) + (x*32)), 1, 32)])
      }
    }
    for (x.outer: int32, 0, 32) "parallel" {
      attr [C.global: handle] "storage_scope" = "global";
      allocate(C.global, float32, [1024]);
      for (y.outer: int32, 0, 32) {
        for (x.c.init: int32, 0, 32) {
          C.global[ramp((x.c.init*32), 1, 32)] = broadcast(0f32, 32)
        }
        for (k.outer: int32, 0, 256) {
          for (x.c: int32, 0, 32) {
            C.global[ramp((x.c*32), 1, 32)] = ((float32x32*)C.global[ramp((x.c*32), 1, 32)]) + (broadcast((float32*)A_2[(((x.outer*32768) + (x.c*1024)) + (k.outer*4))]), 32)*(float32x32*)packedB[ramp(((y.outer*32768) + (k.outer*128)), 1, 32)])))
            C.global[ramp((x.c*32), 1, 32)] = ((float32x32*)C.global[ramp((x.c*32), 1, 32)]) + (broadcast((float32*)A_2[((((x.outer*32768) + (x.c*1024)) + (k.outer*4)) + 1)]), 32)*(float32x32*)packedB[ramp((((y.outer*32768) + (k.outer*128)) + 32), 1, 32)])))
            C.global[ramp((x.c*32), 1, 32)] = ((float32x32*)C.global[ramp((x.c*32), 1, 32)]) + (broadcast((float32*)A_2[((((x.outer*32768) + (x.c*1024)) + (k.outer*4)) + 2)]), 32)*(float32x32*)packedB[ramp((((y.outer*32768) + (k.outer*128)) + 64), 1, 32)])))
            C.global[ramp((x.c*32), 1, 32)] = ((float32x32*)C.global[ramp((x.c*32), 1, 32)]) + (broadcast((float32*)A_2[((((x.outer*32768) + (x.c*1024)) + (k.outer*4)) + 3)]), 32)*(float32x32*)packedB[ramp((((y.outer*32768) + (k.outer*128)) + 96), 1, 32)])))
          }
        }
        for (x.inner: int32, 0, 32) {
          for (y.inner: int32, 0, 32) {
            C_2[((((x.outer*32768) + (x.inner*1024)) + (y.outer*32)) + y.inner)] = (float32*)C.global[((x.inner*32) + y.inner)])
          }
        }
      }
    }
  }
}
```

tqchen commented:
cc @jroesch @antinucleon @junrushao1994 @Hzfengsy @mbrookhart

Hzfengsy commented:
Some comments:
1. There are more than one spaces before the left brace in the allocation line 
    ```
    allocate(B.local, float32, [64])  {
    ```

2. Can we use the same rule for the allocation stmt as the one for attr? Allocation stmt now will bring extra indentation

3. ```
    attr [IterVar(blockIdx.z: int32, [(nullptr)], "ThreadIndex", "blockIdx.z")] "thread_extent" = 196;
    ```
    It is strange to print `nullptr` here especially in square brackets. Perhaps we can use `IterVar(blockIdx.z: int32, (nullptr), "ThreadIndex", "blockIdx.z")]` or `IterVar(blockIdx.z: int32, , "ThreadIndex", "blockIdx.z")]`

4. Considering future parsing use, we must print the dtype for every const number. But we may use some shorthand for common dtype. e.g. `2f` for float32, `2h` for float16(half), direct`2` for int32 (for here most integer numbers in schedule are int32). But still, keep the complete form for every type. e.g. `int8(2)`, `float64(2)`(or may be `fp64(2)`) , also, `float32(2)` is legal as well.

tqchen commented:
- Allocate `float32x32` is not the actual data type. Perhaps we can just show allocate the flattened size since that is the semantics. 
- eval is not necessary since they can be implied in a call. 
- We might need to update call later to something like `@intrin.func_name(args)`

junrushao1994 commented on src/printer/tir_text_printer.cc:
```diff
@@ -0,0 +1,735 @@
+   * \param dtype The data type
+   */
+  static Doc PrintDType(DataType dtype) {
+    return Doc::Text(runtime::DLDataType2String(dtype));
```
Shall we consider printing something non-empty when `dtype.bits() == 0`?

spectrometerHBH replied:
When will we encounter a dtype with bits=0?

junrushao1994 replied:
Something like void. I don’t think that happens very often though

junrushao1994 commented on src/printer/tir_text_printer.cc:
```diff
@@ -0,0 +1,735 @@
+   * \param data The pointer to hold the data.
+   */
+  template <typename T>
+  static Doc PrintConstScalar(DataType dtype, const T* data) {
```
Why not just do like this?

```suggestion
  static Doc PrintConstScalar(DataType dtype, const T& data) {
```

spectrometerHBH replied:
fixed

junrushao1994 commented on src/printer/tir_text_printer.cc:
```diff
@@ -0,0 +1,735 @@
+    std::string unique_prefix = prefix;
+    auto it = name_alloc_map_.find(prefix);
+    if (it != name_alloc_map_.end()) {
+      while (true) {
+        std::ostringstream os;
+        os << prefix << "_" << (++it->second);
+        std::string name = os.str();
+        if (name_alloc_map_.count(name) == 0) {
+          unique_prefix = name;
+          break;
+        }
+      }
```
```suggestion
      while (name_alloc_map_.count(
        unique_prefix =
          prefix + "_" + std::to_string(++it->second)
      ) > 0);
```

spectrometerHBH replied:
fixed

junrushao1994 commented on src/printer/tir_text_printer.cc:
```diff
@@ -0,0 +1,735 @@
+  }
+
+  Doc GetUniqueName(std::string prefix) {
+    // std::replace(prefix.begin(), prefix.end(), '.', '_');
```
Not sure whether we should substitute "." with "_". Probably remove this line if we have some consensus? @tqchen

junrushao1994 commented on src/printer/tir_text_printer.cc:
```diff
@@ -0,0 +1,735 @@
+    if (name.length() == 0 || !std::isalpha(name[0])) {
+      name = "v" + name;
+    }
+    Doc val = GetUniqueName(name);
```
Just to clarify, what's the result of `GetUniqueName(name)` if `name.length() == 0` for the first time, `v` or `v0`?

spectrometerHBH replied:
If `v` has been used, then it will be `v`. Otherwise, it will be `v_1`.

junrushao1994 commented on src/printer/tir_text_printer.cc:
```diff
@@ -0,0 +1,735 @@
+
+Doc TIRTextPrinter::Print(const ObjectRef& node) {
+  if (!node.defined()) return Doc::Text("(nullptr)");
+  if (node.as<StmtNode>()) {
```
```suggestion
  if (node->IsInstance<StmtNode>()) {
```

spectrometerHBH replied:
fixed

junrushao1994 commented on src/printer/tir_text_printer.cc:
```diff
@@ -0,0 +1,735 @@
+    case CallNode::Intrinsic:return "intrin";
+    case CallNode::PureIntrinsic:return "pure_intrin";
+  }
+  return "Unknown";
```
It might be better to just `LOG(FATAL)` in this case?

spectrometerHBH replied:
fixed. But return is kept otherwise it will trigger a warning in compilation

junrushao1994 replied:
Thanks! You may add “throw;” to avoid using return :-)

spectrometerHBH commented:
> Some comments:
> 
> 1. There are more than one spaces before the left brace in the allocation line
>    ```
>    allocate(B.local, float32, [64])  {
>    ```
> 2. Can we use the same rule for the allocation stmt as the one for attr? Allocation stmt now will bring extra indentation
> 3. ```
>     attr [IterVar(blockIdx.z: int32, [(nullptr)], "ThreadIndex", "blockIdx.z")] "thread_extent" = 196;
>    ```
>    
>    
>    It is strange to print `nullptr` here especially in square brackets. Perhaps we can use `IterVar(blockIdx.z: int32, (nullptr), "ThreadIndex", "blockIdx.z")]` or `IterVar(blockIdx.z: int32, , "ThreadIndex", "blockIdx.z")]`
> 4. Considering future parsing use, we must print the dtype for every const number. But we may use some shorthand for common dtype. e.g. `2f` for float32, `2h` for float16(half), direct`2` for int32 (for here most integer numbers in schedule are int32). But still, keep the complete form for every type. e.g. `int8(2)`, `float64(2)`(or may be `fp64(2)`) , also, `float32(2)` is legal as well.

1. fixed
2. fixed. But here we implicitly assume that `Allocate` and `Attr` will have at least one child. Otherwise, for such a scenario
```c++
attr...;
attr...;
attr...;
```
We can not determine whether it is `attr|attr|attr` or `attr(attr)|attr`
3. fixed

spectrometerHBH commented:
To avoid making this page too long, I will edit&update the examples for reference in the top if I change the format.

tqchen commented:
Going to merge it in two days

spectrometerHBH commented:
Relay's type printing requires attr and relayExpr printing, so It is not straightforward to make type printing independent. Meanwhile, relay's attir printing overlaps PrimExpr printing in tir now. We'd better combine them in later PRs.

tqchen commented on python/tvm/tir/parser.py:
```diff
@@ -0,0 +1,24 @@
+from . import _ffi_api
+
+
+def astext(code):
```
Likely we should not put astext here. Because irnode.astext already redirects to astext in the backend? If we hack relay's astext to redirect. Alternatively overload the astext member function of the TIR Nodes (PrimExpr, PrimFunc).

spectrometerHBH replied:
I see. But why tir.Stmt is not a subclass of Node now?

spectrometerHBH replied:
It is a subclass of Object in tir/stmt.py

tqchen replied:
there is not special Node class atm. For now let us just add astext to stmt as well.

xqdan commented:
Binds in some pass function is not clean for round trip dump ir, how do we deal with it?

Hzfengsy commented:
> Binds in some pass function is not clean for round trip dump ir, how do we deal with it?

Do you mean that `buffer_bind` is not necessary to print after the pass `storage_flatten`?
Yes, the buffer no longer exists after the flatten pass. All we have is that Var(buffer->data) in Load and Store. But we still need a place to define those vars, and that's why we still have buffer_bind, where we define buffer as well as buffer->data, in low-level functions

Furthermore, maybe one day we can use BufferLoad/ BufferStore after the flatten and use buffer from the beginning to the end.

spectrometerHBH commented:
I can not reproduce the errors in CI.

spectrometerHBH commented:
I found the problem. It is due to the different behavior of gcc5.4 and gcc 7.4
```c++
doc << "(" << Print(op->a) << OpString << Print(op->b) << ")";
```

gcc5.4 will execute Print(op->b) first.
gcc7.4 will execute Print(op->a) first.

spectrometerHBH commented:
Both `release` and `debug` under gcc5.4 will show the same behavior. It is not likely that O2 causes this problem.

spectrometerHBH commented:
clang6.0 also works. But I will fix it under gcc5.4

spectrometerHBH commented:
Take another look please @tqchen

tqchen commented:
THanks @spectrometerHBH @xqdan @junrushao1994 @Hzfengsy !

xqdan commented:
> > Binds in some pass function is not clean for round trip dump ir, how do we deal with it?
> 
> Do you mean that `buffer_bind` is not necessary to print after the pass `storage_flatten`?
> Yes, the buffer no longer exists after the flatten pass. All we have is that Var(buffer->data) in Load and Store. But we still need a place to define those vars, and that's why we still have buffer_bind, where we define buffer as well as buffer->data, in low-level functions
> 
> Furthermore, maybe one day we can use BufferLoad/ BufferStore after the flatten and use buffer from the beginning to the end.

https://github.com/apache/incubator-tvm/blob/master/src/driver/driver_api.cc#L139

that's i'm asking, seems binds is embedding into ir at the beginning, looks good

roastduck commented on src/printer/tir_text_printer.cc:
```diff
@@ -0,0 +1,597 @@
+Doc TIRTextPrinter::VisitExpr_(const LoadNode* op) {
+  Doc doc;
+  doc << "(" << PrintDType(op->dtype) << "*)"
+      << Print(op->buffer_var) << "[" << Print(op->index) << "])";
```
Parentheses are not matched here. Please consider `((dtype*)buffer_var)[index]`.
